functions/kubernetes/k8sJobSubmit.js

```diff
@@ -176,24 +202,31 @@ var submitK8sJob = async(kubeconfig, job, taskId, context, customParams) => {
     let taskEnd = new Date().toISOString();
     console.log('Job ended with result:', jobResult, 'time:', taskEnd);
     var code = parseInt(jobResult[1]); // job exit code
-    // if job failed and restart policy is enabled, restart the job
-    if (code != 0 && restartPolicy == "OnFailure" && restartCount++ < backoffLimit)  {
-      console.log("Job", taskId, "failed, restarting (restartCount:", restartCount + ") ...");
-      // When restarting, need to send the message job again!
-      try {
-        await context.sendMsgToJob(jobMessage, taskId);
-      } catch (err) {
-        console.error(err);
-        throw err;
-      }
-      return awaitJob(taskId); // just wait again assuming Kubernetes will restart the job
-    } 
     return code;
   }
 
-  let jobExitCode = awaitJob(taskId);
-  return jobExitCode;
+  var awaitJobs = async(taskIdArr) => {
+    let awaitPromises = []
+    for (var i=0; i<taskIdArr.length; i++) {
+      awaitPromises.push(awaitJob(taskIdArr[i]));
+    }
+    return Promise.all(awaitPromises);
+  }
+
+  let jobExitCodes = [];
+  while (true) {
+    jobExitCodes = await awaitJobs(taskIdArr);
+    // we stop if every task exited with 0 or no more restarts are possible
+    if (jobExitCodes.every(item => item == 0) || restartPolicy != "OnFailure" || restartCount++ >= backoffLimit) {
+      break;
+    }
+    console.log("Job", taskId, "failed, restarting (restartCount:", restartCount + ") ...");

```

Comment:
Looks like we no longer have `taskId` here! It looks that job arrays mess up the restart behavior. What SHOULD happen if only one `jobExitCode != 0`? I think the answer is: the job should be buffered again and executed in a different job group with incremented `restartCount`. (Note that the old restart mechanism relied on Kubernetes restarting the **Pod**. We can no longer do this since it would re-execute the entire job group.) I suspect this cannot be easily fixed, so I propose to add a FIXME comment here.

Reply:
Fixed: added `RestartCounter` class and `restartFn` that is used to rebuffer failed tasks.